### ISSUE TITLE
Switchbot Password & Switchbot Homeassistant State

### ISSUE THREAD

**Issue by Migz93** [fphammerle/switchbot-mqtt#37]
Thanks for this tool, it's working a treat for a switchbot that is just a little too far away from home assistant and used to only sporadically work.

I'm not sure if I missed something but is it possible to keep the password on my switchbot device to use with this? I couldn't find anywhere to specify the password so I had to remove it completely. Worried that anybody around me could get the switchbot app and activate that switchbot, I know it's fairly unlikely.

I was also wondering if as I have my switchbot set in press mode, if it's possible that when I turn it "on" in home assistant, it shortly afterwards sets itself back to off? I could set an automation to automatically set it back to "off" after being set to "on" but then that would activate the switchbot again which I don't want.

**Comment by fphammerle** [comment]
Hi,

> is it possible to keep the password on my switchbot device to use with this?

support for passwords has not yet been implemented, but it's on my todo list.
I am thinking about adding support for a config file with a mac address to password mapping, e.g.:
```yaml
# switchbot-passwords.yml
aa:bb:cc:dd:ee:ff: secret
dd:ee:ff:aa:bb:cc: password
```

> I was also wondering if as I have my switchbot set in press mode, if it's possible that when I turn it "on" in home assistant, it shortly afterwards sets itself back to off?

Could you use a `script` instead of `button`?

```yaml
# home assistant config
script:
  trigger_some_switchbot:
    sequence:
    - service: mqtt.publish
      data:
        topic: homeassistant/switch/switchbot/aa:bb:cc:dd:ee:ff/state
        payload: ON
 ```

**Comment by Migz93** [comment]
Thanks for the quick reply!

`I am thinking about adding support for a config file with a mac address to password mapping, e.g.:`
Yeah I'd be more then happy with a config file to map like that. If that's something you're planning to do that's great, thank you.

`Could you use a script instead of button?`
Sorry I was a little unclear. I didn't mean so much for activating the switchbot, as I'm happy enough just using a toggle wherever I need to.
I meant just more in the way this communicates with home assistant, if switchbot-mqtt can read the mode of a switchbot and if the mode is set to press mode then whenever the switch is activated, automatically set it back to off a after a few seconds. 
It's not a big one to me as it works completely fine just using toggle, I just think maybe when it's in press mode it's more accurate to have it set itself back to off as that's sort of what the switchbot physically does.

**Comment by fphammerle** [comment]
> if switchbot-mqtt can read the mode of a switchbot and if the mode is set to press mode then whenever the switch is activated, automatically set it back to off a after a few seconds.

Currently, `switchbot-mqtt` only reports on home-assistant's `state_topic` **once** after the switchbot accepted a command, even in press mode.
In press mode, as you said, `switchbot-mqtt` should actually report to `state_topic` **twice** (first `ON`, then after a few seconds `OFF`).

The underlying library `pySwitchbot` recently added support for fetching the button's mode of operation: https://github.com/Danielhiversen/pySwitchbot/pull/22/files#diff-4a04772ec76f55c7c44a96ff0eef45da07305d1e6ef7100e3a3c6d2a80f32e35R222 (release: https://github.com/Danielhiversen/pySwitchbot/blob/0.11.0/switchbot/__init__.py#L398)

So, `switchbot-mqtt`'s state reporting behavior should be adapted to honor `Switchbot.switch_mode` (switch or press mode).

Thanks for reporting that issue!

**Comment by fphammerle** [comment]
`Switchbot.switch_mode` (https://github.com/Danielhiversen/pySwitchbot/blob/0.11.0/switchbot/__init__.py#L398) currently requires a prior call to `SwitchbotDevice.get_device_data` (https://github.com/Danielhiversen/pySwitchbot/blob/0.11.0/switchbot/__init__.py#L309).

Unfortunately, in the current docker setup these calls fail due to insufficient permissions.
Please see https://github.com/fphammerle/switchbot-mqtt/pull/31#issuecomment-846383603 for details.

In summary, calling `SwitchbotDevice.get_device_data` would trigger `bluepy.btle.BTLEManagementError: Failed to execute management command 'le on' (code: 20, error: Permission Denied)`.
I haven't found a way yet to configure the linux kernel to allow these commands (while keeping the `no-new-privileges` flag).

**Comment by fphammerle** [comment]
New release supports password-protected devices: https://github.com/fphammerle/switchbot-mqtt/tree/v1.0.0#device-passwords

**Comment by Migz93** [comment]
Thanks for implementing this!
Had some issues initially but I think it's because the docker image isn't compatible with my RPI0W and for some reason I couldn't use the dockerfile directly.
I've just swapped to using the pip3 switchbot-mqtt command instead and it's working perfectly.

**Comment by fphammerle** [comment]
Sorry, I had not pushed the `armv6` image for RPI0 and RPI1 yet.

Could you retry with the new image `docker.io/fphammerle/switchbot-mqtt:1.0.0-armv6` ?

From my experience, `docker build` on raspberry pi zeros typically fails because the docker daemon pulls the base image for the wrong architecture.
Workaround: pull base image before `docker build` and manually specify architecture: `sudo docker pull --platform linux/arm/v6 docker.io/alpine:3.13.1`

**Comment by Migz93** [comment]
Ahah that explains it.
Tried with the new image but I'm getting
```
Fatal Python error: pyinit_main: can't initialize time
Python runtime state: core initialized
PermissionError: [Errno 1] Operation not permitted

Current thread 0xb6f31390 (most recent call first):
<no Python frame>
```

I initially tried just updating the image via Portainer but when that failed I deleted everything and ran the following command
`docker run --name switchbot --userns host --network host "fphammerle/switchbot-mqtt:1.0.0-armv6" switchbot-mqtt --mqtt-host 192.168.12.2`
I intentionally left out anything related to the device passwords just to rule out that being an issue.

Thank you for explaining that workaround, interesting to know and will keep it for future reference.

**Comment by fphammerle** [comment]
I think the `PermissionError` you got is caused by an incompatibility of the alpine base image with the `libseccomp` installed on the base system (I assume it's Raspbian 10/Buster):
https://blog.samcater.com/fix-workaround-rpi4-docker-libseccomp2-docker-20/

I hope the new Raspbian / Raspberry Pi OS version will be released soon (11/Bullseye).
Then this bug should be fixed.

In the meantime, adding `--security-opt seccomp=unconfined` to `docker run` might fix the issue, but on the downside that option is disabling some of docker's security feaetures.

**Comment by Migz93** [comment]
I'm happy to forego a bit of security until the new release of raspbian, especially as this RPI only runs switchbot-mqtt.
Running it with that extra security-opt command did the trick. Thank you for the help and thanks for putting the device password in!

**Comment by fphammerle** [comment]
Pre-release [v4.0.0a0](https://pypi.org/project/switchbot-mqtt/4.0.0a0/#history) replaced bluepy with bleak.

In my tests, fetching basic device info (battery status etc.) no longer requires any capapbilities or security opts.
